src/transformer.ts

```diff
@@ -291,53 +284,131 @@ export default class Transformer {
+       */
+      const fromPath = path.join(Transformer.outputPath, 'schemas', 'objects');
+      const toPath = Transformer.prismaClientOutputPath!;
+      const relativePathFromOutputToPrismaClient = path
+        .relative(fromPath, toPath)
+        .split(path.sep)
+        .join(path.posix.sep);
+      prismaClientImportPath = relativePathFromOutputToPrismaClient;
+    } else {
+      /**
+       * If the default output path for prisma client (@prisma/client) is being used, we can import from it directly
+       * without having to resolve a relative path
+       */
+      prismaClientImportPath = Transformer.prismaClientOutputPath;
+    }
+    return `import type { Prisma } from '${prismaClientImportPath}';\n\n`;
+  }
 
-    const prismaImport = this.getImportPrisma();
+  generateJsonSchemaImplementation() {
+    let jsonSchemaImplementation = '';
 
-    const json = this.getJsonSchemaImplementation();
+    if (this.hasJson) {
+      jsonSchemaImplementation += `\n`;
+      jsonSchemaImplementation += `const literalSchema = z.union([z.string(), z.number(), z.boolean()]);\n`;
+      jsonSchemaImplementation += `const jsonSchema: z.ZodType<Prisma.InputJsonValue> = z.lazy(() =>\n`;
+      jsonSchemaImplementation += `  z.union([literalSchema, z.array(jsonSchema.nullable()), z.record(jsonSchema.nullable())])\n`;
+      jsonSchemaImplementation += `);\n\n`;
+    }
 
-    return `${this.getImportsForObjectSchemas()}${prismaImport}${json}${objectSchema}`;
+    return jsonSchemaImplementation;
   }
 
-  async printObjectSchemas() {
-    const zodStringFields = this.fields
-      .map((field) => this.getObjectSchemaLine(field))
-      .flatMap((item) => item)
-      .map((item) => {
-        const [zodStringWithMainType, field, skipValidators] = item;
+  generateObjectSchemaImportStatements() {
+    let generatedImports = this.generateImportZodStatement();
+    generatedImports += this.generateSchemaImports();
+    generatedImports += '\n\n';
+    return generatedImports;
+  }
 
-        const value = skipValidators
-          ? zodStringWithMainType
-          : this.getFieldValidators(zodStringWithMainType, field);
+  generateSchemaImports() {
+    return [...this.schemaImports]
+      .map((name) => {
+        const { isModelQueryType, modelName, queryName } =
+          this.checkIsModelQueryType(name);
+        if (isModelQueryType) {
+          return `import { ${this.resolveModelQuerySchemaName(
+            modelName!,
+            queryName!,
+          )} } from '../${queryName}${modelName}.schema'`;
+        } else if (Transformer.enumNames.includes(name)) {
+          return `import { ${name}Schema } from '../enums/${name}.schema'`;
+        } else {
+          return `import { ${name}ObjectSchema } from './${name}.schema'`;
+        }
+      })
+      .join(';\r\n');
+  }
 
-        return value.trim();
-      });
+  checkIsModelQueryType(type: string) {
```

Comment:
This function is used in `generateSchemaImports`, and the cases there only `findMany`, `enum` and `object schemas`. My question is how does it handle others like findUnique?

I might be missing something about `checkIsModelQueryType`.

Reply:
Great question! First I will try to write out some background info for useful context:

When generating `{Model}Select` in `generateModelSelectInputObjectTypes`, we have the following line that is inspired by the generated Prisma types for select:

`type: isList ? `${type}FindManyArgs` : `${type}Args`,`

Examples:
On UserSelect: `posts: boolean | PostFindManyArgs`
On PostSelect: `author: boolean | UserArgs`

For`{type}Args`, this is a new zod schema we will have to generate, which we account for in `generateModelArgsInputObjectTypes`

However for `{type}FindManyArgs`, our `{Model}FindManySchema` at `findMany{Model}.ts` is the perfect fit, and we do not have to generate another file.

The tricky part is that up until now `generateSchemaImports` did not need to support importing our top level zod query schemas, which is why `checkIsModelQueryType` and `resolveModelQuerySchemaName` came into existence, and why we `import { ${name}Schema } from '../enums/${name}.schema'` 

Currently we only have the `FindManyArgs -> 'findMany'` mapping in `checkIsModelQueryType`, as this is the first time that: one of the fields in an input type object we've pushed is a top level query schema (findMany{Model}). 

And lastly now that we've generated the `{Model}Select` schemas, we can add `select` to the top level query zod schemas that support select in `generateModelSchemas`. `z.object({ select: ${modelName}SelectObjectSchema.optional(),` 🙂

Reply:
@BitPhoenix Thank you for the detailed explanation, really drove the point home.